sw/device/lib/dif/dif_spi_host.c

```diff
@@ -0,0 +1,382 @@
+    ready = bitfield_bit32_read(reg, SPI_HOST_STATUS_READY_BIT);
+  } while (!ready);
+}
+
+static void wait_tx_fifo(const dif_spi_host_t *spi_host) {
+  uint32_t txqd;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    txqd = bitfield_field32_read(reg, SPI_HOST_STATUS_TXQD_FIELD);
+  } while (txqd == SPI_HOST_PARAM_TX_DEPTH);
+}
+
+static void wait_rx_fifo(const dif_spi_host_t *spi_host) {
+  uint32_t rxqd;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    rxqd = bitfield_field32_read(reg, SPI_HOST_STATUS_RXQD_FIELD);
+  } while (rxqd == 0);
+}
+
+static inline void tx_fifo_write8(const dif_spi_host_t *spi_host,
+                                  uintptr_t srcaddr) {
+  uint8_t *src = (uint8_t *)srcaddr;
+  wait_tx_fifo(spi_host);
+  mmio_region_write8(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET, *src);
+}
+
+static inline void tx_fifo_write32(const dif_spi_host_t *spi_host,
+                                   uintptr_t srcaddr) {
+  uint32_t *src = (uint32_t *)srcaddr;
+  wait_tx_fifo(spi_host);
+  mmio_region_write32(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET, *src);
+}
+
+void dif_spi_host_fifo_write(const dif_spi_host_t *spi_host, const void *src,
+                             uint16_t len) {
+  uintptr_t ptr = (uintptr_t)src;
+
+  // If the pointer starts mis-aligned, write until we are aligned.
+  while (misalignment32_of(ptr) && len > 0) {
+    tx_fifo_write8(spi_host, ptr);
+    ptr += 1;
+    len -= 1;
+  }
+
+  // Write complete 32-bit words to the fifo.
+  while (len > 3) {
+    tx_fifo_write32(spi_host, ptr);
+    ptr += 4;
+    len -= 4;
+  }
+
+  // Clean up any leftover bytes.
+  while (len > 0) {
+    tx_fifo_write8(spi_host, ptr);
+    ptr += 1;
+    len -= 1;
+  }
+}
+
+typedef struct queue {
+  int32_t length;
+  uint8_t alignas(uint64_t) data[8];
+} queue_t;
+
+static void enqueue_byte(queue_t *queue, uint8_t data) {
+  queue->data[queue->length++] = data;
+}
+
+static void enqueue_word(queue_t *queue, uint32_t data) {
+  if (queue->length % sizeof(uint32_t) == 0) {
+    write_32(data, queue->data + queue->length);
+    queue->length += 4;
+  } else {
+    enqueue_byte(queue, (uint8_t)data);
+    data >>= 8;
+    enqueue_byte(queue, (uint8_t)data);
+    data >>= 8;
+    enqueue_byte(queue, (uint8_t)data);
+    data >>= 8;
+    enqueue_byte(queue, (uint8_t)data);
+  }
+}
+
+static uint8_t dequeue_byte(queue_t *queue) {
+  uint8_t val = queue->data[0];
+  uint64_t qword = read_64(queue->data);
+  write_64(qword >> 8, queue->data);
+  queue->length -= 1;
+  return val;
+}
+
+static uint32_t dequeue_word(queue_t *queue) {
+  uint32_t val = read_32(queue->data);
+  write_32(read_32(queue->data + sizeof(uint32_t)), queue->data);
+  queue->length -= 4;
+  return val;
+}
+
+void dif_spi_host_fifo_read(const dif_spi_host_t *spi_host, void *dst,
```

Comment:
I think that the functions `dif_spi_host_fifo_read` and `dif_spi_host_fifo_read` should return `dif_result_t` and verify the imput arguments before anything else.

Reply:
There are no bad input parameters: there are no requirements on src/dst pointer alignment, nor on the length value.